Tactics/SymContext.lean

```diff
@@ -132,57 +148,87 @@ def findFromUserName (name : Name) : MetaM LocalDecl := do
+def getHRunName : m Name := do return (← read).h_run
+
+/-- Retrieve the name for the next state
+
+NOTE: `getNextStateName` does not increment the state, so consecutive calls
+will give the same name. Calling `prepareForNextStep` will increment the state.
+-/
+def getNextStateName : m Name := do
+  let c ← read
+  return Name.mkSimple s!"{c.state_prefix}{c.currentStateNumber + 1}"
+
+end Monad
+
 end
 
-/-! ## `ToMessageData` instance -/
+/-! ## `ToMessageData` instance and tracing -/
 
 /-- Convert a `SymContext` to `MessageData` for tracing.
 This is not a `ToMessageData` instance because we need access to `MetaM` -/
 def toMessageData (c : SymContext) : MetaM MessageData := do
-  let state ← c.stateExpr
   let h_run ← userNameToMessageData c.h_run
-  let h_err? ← c.h_err?.mapM userNameToMessageData
   let h_sp?  ← c.h_sp?.mapM userNameToMessageData
 
-  return m!"\{ state := {state},
-  finalState := {c.finalState},
+  return m!"\{ finalState := {c.finalState},
   runSteps? := {c.runSteps?},
   h_run := {h_run},
   program := {c.program},
   pc := {c.pc},
-  h_err? := {h_err?},
   h_sp? := {h_sp?},
   state_prefix := {c.state_prefix},
-  curr_state_number := {c.curr_state_number} }"
+  curr_state_number := {c.currentStateNumber},
+  effects := {c.effects} }"
+
+variable {α : Type} {m : Type → Type} [Monad m] [MonadTrace m] [MonadLiftT IO m]
+  [MonadRef m] [AddMessageContext m] [MonadOptions m] {ε : Type}
+  [MonadAlwaysExcept ε m] [MonadLiftT BaseIO m] in
+def withSymTraceNode (msg : MessageData) (k : m α) : m α := do
+  withTraceNode `Tactic.sym (fun _ => pure msg) k
+
+def traceSymContext : SymM Unit :=
+  withTraceNode `Tactic.sym (fun _ => pure m!"SymContext: ") <| do
+    let m ← (← getThe SymContext).toMessageData
+    trace[Tactic.sym] m
 
 /-! ## Creating initial contexts -/
 
+/-- Modify a `SymContext` with a monadic action `k : SymM Unit` -/
+def modify (ctxt : SymContext) (k : SymM Unit) : TacticM SymContext := do
+  let ((), ctxt) ← SymM.run ctxt k
+  return ctxt
+
 /-- Infer `state_prefix` and `curr_state_number` from the `state` name
 as follows: if `state` is `s{i}` for some number `i` and a single character `s`,
 then `s` is the prefix and `i` the number,
 otherwise ignore `state`, and start counting from `s1` -/
-def inferStatePrefixAndNumber (ctxt : SymContext) : SymContext :=
-  let state := ctxt.state.toString
+def inferStatePrefixAndNumber : SymM Unit := do
+  let state ← AxEffects.getCurrentStateName
+  let state := state.toString
   let tail := state.toSubstring.drop 1
 
-  if let some curr_state_number := tail.toNat? then
-    { ctxt with
-      state_prefix := (state.get? 0).getD 's' |>.toString,
-      curr_state_number }
-  else
-    { ctxt with
-      state_prefix := "s",
-      curr_state_number := 1 }
+  modifyThe SymContext fun ctxt =>
+    if let some currentStateNumber := tail.toNat? then
+      { ctxt with
+        state_prefix := (state.get? 0).getD 's' |>.toString,
+        currentStateNumber }
+    else
```

Comment:
Let's print a warning here that tells the user that `s1` is the state var we're running with

Reply:
Sure, I've added a warning